### PR TITLE
Cleanup dtype mechanism

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Adding a new dtype was somewhat tedious and required modification of `scipp::core`, adding the new type to and `enum`. Furthermore, it was necessary to define string formatting. This is an attempt to improve this.

- Using RTTI to define `dtype`.
- Display names (and names of Python dtype attributes) defined via new template argument in `INSTANTIATE_VARIABLE`.
- No significant slowdown observable in the (presumably) most sensitive benchmark, `BM_groupby_large_table`.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
@reviewers: Please review this carefully, the RTTI and static stuff is not the thing I am most familiar with.

File: variable/include/scipp/variable/variable.tcc [hunk 1]
```diff
@@ -785,7 +785,11 @@ Variable Variable::create(const units::Unit &u, const Dims &d, const Shape &s,
 
 /// Macro for instantiating classes and functions required for support a new
 /// dtype in Variable.
-#define INSTANTIATE_VARIABLE(...)                                              \
+#define INSTANTIATE_VARIABLE(name, ...)                                        \
```
Comment (OwenArnold) [inline variable/include/scipp/variable/variable.tcc:788]: Why variadic? Shouldn't we only have a single second argument to denote a single type

File: core/include/scipp/core/dtype.h [hunk 2]
```diff
@@ -32,71 +23,13 @@ template <class T>
...
-  SpanFloat,
-  SpanInt64,
-  SpanInt32,
-  SpanConstDouble,
-  SpanConstFloat,
-  SpanConstInt64,
-  SpanConstInt32,
-  Unknown
+struct SCIPP_CORE_EXPORT DType {
+  std::type_index index;
```
Comment (OwenArnold) [inline core/include/scipp/core/dtype.h:27]: should be const i think. Can't see any reason we would want this to change.

File: variable/include/scipp/variable/variable.tcc (hunk 1 above)
Comment (SimonHeybrock) [reply variable/include/scipp/variable/variable.tcc:788]: Templates such as `std::pair` that contain a comma are recognizedas two arguments by the preprocessor. We don't have any of those right now, but we did in the past.

File: core/include/scipp/core/dtype.h (hunk 2 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/dtype.h:27]: We need `DType` to be assignable and copyable. Relevant classes like `Variable` never expose a mutable reference to their `DType`, so changing it does not harm.

**Pull request merged**